man/systemd-stub.xml

```diff
@@ -71,8 +71,12 @@
       <listitem><para>A boot splash (in Windows <filename>.BMP</filename> format) to show on screen before
       invoking the kernel will be looked for in the <literal>.splash</literal> PE section.</para></listitem>
 
-      <listitem><para>A compiled binary DeviceTree will be looked for in the <literal>.dtb</literal> PE
-      section.</para></listitem>
+      <listitem><para>Compiled binary DeviceTrees will be looked for in the <literal>.dtb</literal> PE
+      sections. The stub will try to match the DeviceTrees embedded in the PE image with the DeviceTree
+      provided by the device's firmware by comparing the first <literal>compatible</literal> string in each UKI DeviceTree
+      with the first <literal>compatible</literal> string supplied by the firmware. If no match is found the first
+      <literal>.dtb</literal> section in the image is selected.
```

Comment:
> If no match is found the first `.dtb` section in the image is selected.

worries me. as that will break booting the same uki in a QEMU KVM and hardware. Because one should not override the QEMU KVM provided DTB.

I am also worried about systems that are pure ACPI and thus do not have DeviceTree provided by the firmware -> because then we need to probably extend dtb with things to match on, and likely match on like SMBIOS3 table vendor / laptop SKU. This is to support all the Windows On Arm laptops. 

I wonder if we can change this to:

> If UKI has only one dtb, it will always be loaded. If UKI has more than one dtb, and none of them match, none of them will be loaded.

As that is both backwards compatible, and requires strict match out of multiple dtbs.

Reply:
And with no dtb loaded, WoA devices would just not work. So how about  something similar to what [dtbloader](https://github.com/TravMurav/dtbloader) does?
stub will have a list of hwids and compatibles, either in the source code or as in separate sections of the resulting .efi binary. While the second option isn't in the standard, but at least it'll make having more than one dtb useful even without recompiling the stub, just with a configuration file. And nothing prevents this from making it to the standard some day.

Or, even better, both `.hwid` sections and hardcoded list will be used, and all `.hwid` sections will be checked before the hardcoded list.

So the description will be the loosely the following:
> If UKI has only one dtb, it will always be loaded. If UKI has more than one dtb, none of them match firmware-defined dtb, but device's hwid is present either in `compatible_devices_list` (as defined in <reference to the source file>) or in one of the `.hwid` sections, the corresponding dtb will be loaded. In any other case, none of the dtbs will be loaded.